engine/resources/hetzner_vm.go

```diff
@@ -0,0 +1,804 @@
+// serverconfig in hcloud-go format from resource params, requests a server
+// creation with that configuration, and waits until the creation is confirmed.
+// Errors occur when the server exists already, the client fails, or the wait
+// step exits early (context cancellation, client failure or timeout).
+// TODO: add createServer permission field
+func (obj *HetznerVMRes) createServer(ctx context.Context) error {
+	if obj.init.Debug {
+		obj.init.Logf("createServer()")
+	}
+	if obj.server != nil {
+		return fmt.Errorf("server already exists")
+	}
+	if err := obj.getServerConfig(ctx); err != nil {
+		return errwrap.Wrapf(err, "getServerConfig failed")
+	}
+	if _, _, err := obj.client.Server.Create(ctx, obj.serverconfig); err != nil {
+		return errwrap.Wrapf(err, "client.server.create failed")
+	}
+	if err := obj.waitFor(ctx, serverExists); err != nil {
+		return errwrap.Wrapf(err, "waitFor(serverExists) exited early")
+	}
+	return nil
+}
+
+// deleteServer checks if the server is available from the client, requests a
+// server deletion from the API, waits for confirmation, then returns. An error
+// is thrown when the server is already unavailable or something fails. Context
+// cancellation allows a clean exit when needed.
+// NOTE: this is a destructive operation. Check permissions before applying!
+// TODO: move the AllowDeletion policy inside this function for safety?
+func (obj *HetznerVMRes) deleteServer(ctx context.Context) error {
+	if obj.init.Debug {
+		obj.init.Logf("deleteServer()")
+	}
+	if obj.server == nil {
+		return fmt.Errorf("server is already unavailable")
+	}
+	if _, err := obj.client.Server.Delete(ctx, obj.server); err != nil {
+		return errwrap.Wrapf(err, "client.server.delete failed")
+	}
+	if err := obj.waitFor(ctx, serverAbsent); err != nil {
+		return errwrap.Wrapf(err, "waitFor(serverAbsent) exited early")
+	}
+	return nil
+}
+
+// rebuildServer deletes the current server instance and creates a new one, in
+// accordance with the provided resource specifications.
+// NOTE: this is a destructive operation, don't call without AllowDeletion.
+func (obj *HetznerVMRes) rebuildServer(ctx context.Context) error {
+	if obj.init.Debug {
+		obj.init.Logf("rebuildServer()")
+	}
+	if obj.server == nil {
+		return fmt.Errorf("server is unavailable")
+	}
+	if err := obj.deleteServer(ctx); err != nil {
+		return errwrap.Wrapf(err, "deleteServer failed")
+	}
+	if err := obj.createServer(ctx); err != nil {
+		return errwrap.Wrapf(err, "createServer failed")
+	}
+	return nil
+}
+
+// getServerConfig converts resource fields to hcloud.ServerCreateOpts struct,
+// such that a server instance created with those Opts has the desired specs.
+func (obj *HetznerVMRes) getServerConfig(ctx context.Context) error {
+	if obj.init.Debug {
+		obj.init.Logf("getServerConfig()")
+	}
+	// default, volumes not supported (yet)
+	automount := false
+	// poweron at creation to reach `running` faster
+	startAfterCreate := (obj.State == StateRunning)
```

Comment:
set a default and then:
xxx := default
`if obj.State == whatever {
  xxx = blah
}